zebra-state/src/service/check.rs

```diff
@@ -391,19 +413,72 @@ where
 
 /// Perform initial contextual validity checks for the configured network,
 /// based on the committed finalized and non-finalized state.
+/// The semantically verified block is modified here to include any deferred pool balance change.
 ///
 /// Additional contextual validity checks are performed by the non-finalized [`Chain`].
 pub(crate) fn initial_contextual_validity(
     finalized_state: &ZebraDb,
     non_finalized_state: &NonFinalizedState,
-    semantically_verified: &SemanticallyVerifiedBlock,
+    semantically_verified: &mut SemanticallyVerifiedBlock,
 ) -> Result<(), ValidateContextError> {
     let relevant_chain = any_ancestor_blocks(
         non_finalized_state,
         finalized_state,
         semantically_verified.block.header.previous_block_hash,
     );
 
+    // [ZIP-234]: Block subsidy can be calculated only when the pool balances from previous block are known
+    let network = &non_finalized_state.network;
+    if semantically_verified.height > network.slow_start_interval() {
+        #[cfg(not(zcash_unstable = "zip234"))]
+        let expected_block_subsidy = block_subsidy_pre_nsm(semantically_verified.height, network)?;
+
+        #[cfg(zcash_unstable = "zip234")]
+        let expected_block_subsidy = {
+            let pool_value_balance = non_finalized_state
+                .best_chain()
+                .map(|chain| chain.chain_value_pools)
+                .or_else(|| {
+                    finalized_state
+                        .finalized_tip_height()
+                        .filter(|x| (*x + 1).unwrap() == semantically_verified.height)
+                        .map(|_| finalized_state.finalized_value_pool())
+                });
+
+            let money_reserve = if semantically_verified.height > 1.try_into().unwrap() {
+                pool_value_balance
+                    .expect("a chain must contain valid pool value balance")
+                    .money_reserve()
+            } else {
+                MAX_MONEY.try_into().unwrap()
+            };
+            block_subsidy(semantically_verified.height, network, money_reserve)?
+        };
+
+        // See [ZIP-1015](https://zips.z.cash/zip-1015).
+        semantically_verified.deferred_pool_balance_change = Some(subsidy_is_valid(
+            &semantically_verified.block,
+            network,
+            expected_block_subsidy,
+        )?);
+
+        let coinbase_tx = coinbase_is_first(&semantically_verified.block)
+            .map_err(|e| SubsidyError::Other(format!("invalid coinbase transaction: {e}")))?;
```

Comment:
Consider adding a more specific error case, e.g. `CoinbaseNotFirst`.

Reply:
Fixed! I replaced all cases of `SubsidyError::Other` with more specific errors. Thanks for the suggestion!